CV2-4386 initial ideas for pushing to async

DGaffney opened the pull request:
## Description

Shifts the main routine (only the main routine, not the other uses) of alegre fingerprinting to use async instead of sync. Since this is happening before the merge for video, only image and audio are effected by this change (we'll incorporate video after this though). We send async requests and receive responses from alegre via alegre_webhook.rb, which then calls back to the alegre module to store the results. We have to store results for confirmed and suggested async queries temporarily in redis, and then run the relating step (note that in `get_similar_items_v2_callback` we only return relations once we are confirmed to have every in-process async request returned in redis).

References: CV2-4386

## How has this been tested?

Extensively tested locally and confirmed to be working. A few missing lines of coverage that I'll address once review is complete

## Things to pay attention to during code review

Lots of new functions in alegre_v2.rb - not a ton of editing in place, so pay attention to the new functions/routines.

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have added unit and feature tests, if the PR implements a new feature or otherwise would benefit from additional testing
- [ ] I have added regression tests, if the PR fixes a bug
- [ ] I have added logging, exception reporting, and custom tracing with any additional information required for debugging
- [ ] I considered secure coding practices when writing this code. Any security concerns are noted above.
- [ ] I have commented my code in hard-to-understand areas, if any
- [ ] I have made needed changes to the README
- [ ] My changes generate no new warnings
- [ ] If I added a third party module, I included a rationale for doing so and followed our current [guidelines](https://meedan.atlassian.net/wiki/spaces/ENG/overview#Choose-the-%E2%80%9Cright%E2%80%9D-3rd-party-module)

caiosba commented on app/models/concerns/alegre_v2.rb:
```diff
@@ -243,6 +253,17 @@ def safe_get_sync(project_media, field, params={})
 
+    def cache_items_via_callback(project_media, field, confirmed, results)
+      relationship_type = confirmed ? Relationship.confirmed_type : Relationship.suggested_type
+      type = Bot::Alegre.get_type(project_media)
```
Doesn't look like you're using `type` for anything here, Devin.

caiosba commented on app/models/concerns/alegre_webhooks.rb:
```diff
@@ -12,12 +12,19 @@ def valid_request?(request)
 
     def webhook(request)
       begin
+        puts request.inspect
```
Please remove this before merging (or use a proper `Rails.logger.info` call)

caiosba commented on test/models/bot/alegre_3_test.rb:
```diff
@@ -127,7 +130,7 @@ def teardown
   end
 
-  test "should auto transcribe audio" do
+  test "should auto transcribe audio zzz" do
```
Please remove the :zzz: before merging lol

caiosba reviewed (changes requested):
Thanks Devin! I added some early review comments

caiosba commented on app/models/concerns/alegre_v2.rb:
```diff
@@ -274,15 +308,72 @@ def get_similar_items_v2(project_media, field)
+      self.add_relationships(project_media, self.get_similar_items_v2_callback(project_media, field)) unless project_media.is_blank?
+    end
+
     def get_items_with_similar_media_v2(media_url, threshold, team_ids, type)
       alegre_path = ['audio', 'image'].include?(type) ? self.sync_path_for_type(type) : "/#{type}/similarity/search/"
       # FIXME: Stop using this method from v1 once all media types are supported by v2
       # FIXME: Alegre crashes if `media_url` was already requested before, this is why I append a hash
       self.get_items_with_similar_media("#{media_url}?hash=#{SecureRandom.hex}", threshold, team_ids, alegre_path)
     end
```
@DGaffney, can this method still work as sync for now? I'm thinking about two use cases beyond regular similarity matching that would require further changes to the workflow: tipline search and feed search.

caiosba reviewed (commented):
Code looks clean, thanks Devin for addressing the comments. I left one question there. We can handle that in a separate ticket once this one is done, if it still works. Let's discuss today in the ML call, but I suggest you also test the tipline use case locally. I'll follow up with you on this.

caiosba commented:
Also: As usual, let's make sure integration tests pass for this branch before we merge this :)

DGaffney commented:
@caiosba this is ready for final review!

caiosba commented on app/models/concerns/smooch_search.rb:
```diff
@@ -83,8 +83,14 @@ def is_a_valid_search_result(pm)
     end
 
+    def reject_temporary_results(results)
+      results.select do |result_id, result_data|
```
@DGaffney you're not using `result_id` for anything in that block, is it expected? If so, please rename it to `_result_id`, as per the Code Climate suggestion above.

DGaffney replied:
on it!

caiosba commented on app/models/concerns/alegre_v2.rb:
```diff
@@ -1,4 +1,41 @@
 require 'active_support/concern'
+class TemporaryProjectMedia
+  attr_accessor :team_id, :id, :url, :type, :media
+  def media
```
@DGaffney , since `media` is defined as an `attr_accessor`, this means that methods `media` (getter), `media=` (setter) are created... but then you define a method called `media` (like a custom getter). If you just want the setter, please use `attr_writer` for `:media` instead of `attr_accessor`.

caiosba commented on app/models/concerns/alegre_v2.rb:
```diff
@@ -1,4 +1,41 @@
+      "video" => "UploadedVideo",
+      "audio" => "UploadedAudio",
+    }
+    return Struct.new(:type).new(media_type_map[type])
+  end
+
+  def is_blank?
+    return self.type == "blank"
+  end
+
+  def is_link?
+    return self.type == "link"
+  end
+
+  def is_text?
+    return self.type == "text"
+  end
+
+  def is_image?
+    return self.type == "image"
+  end
+
+  def is_video?
+    return self.type == "video"
+  end
+
+  def is_audio?
+    return self.type == "audio"
+  end
```
Suggestion / non-blocker: You could omit all those `return` words.

caiosba commented on app/models/bot/alegre.rb:
```diff
@@ -398,7 +399,11 @@ def self.transcribe_audio(pm)
   def self.media_file_url(pm)
     # FIXME Ugly hack to get a usable URL in docker-compose development environment.
-    url = (ENV['RAILS_ENV'] != 'development' ? pm.media.file.file.public_url : "#{CheckConfig.get('storage_endpoint')}/#{CheckConfig.get('storage_bucket')}/#{pm.media.file.file.path}")
+    if pm.class == TemporaryProjectMedia
```
Suggestion / non-blocker: You could rewrite as `pm.is_a?(TemporaryProjectMedia)`.

caiosba commented on app/models/concerns/alegre_v2.rb:
```diff
@@ -164,6 +201,7 @@ def get_context(project_media, field=nil)
         has_custom_id: true
       }
       context[:field] = field if field && is_not_generic_field(field)
+      context[:temporary_media] = project_media.class == TemporaryProjectMedia
```
Suggestion / non-blocker: You could rewrite as `pm.is_a?(TemporaryProjectMedia)`.

caiosba commented on app/models/concerns/alegre_v2.rb:
```diff
@@ -263,26 +327,133 @@ def get_confirmed_items(project_media, field)
+
+    def get_cached_data(required_keys)
+      redis = Redis.new(REDIS_CONFIG)
+      Hash[required_keys.collect{|k,v| [k, (Hash[YAML.load(redis.get(v)).collect{|kk,vv| [kk.to_i, vv]}] rescue [])]}]
```
Can you please just add a comment explaining what this line is doing? A lot of Ruby magic here hehe.

caiosba commented on app/models/concerns/alegre_webhooks.rb:
```diff
@@ -13,11 +13,17 @@ def valid_request?(request)
+          redis = Redis.new(REDIS_CONFIG)
+          key = "alegre:webhook:#{doc_id}"
+          redis.lpush(key, request.params.to_json)
+          redis.expire(key, 1.day.to_i)
```
I suggest we add the `redis.expire` call under an `ensure` block... this way, even if something goes wrong, we make sure that Redis is going to be cleared (in that case we may need to declare `redis` outside the begin/rescue block).

caiosba commented on app/models/concerns/smooch_search.rb:
```diff
@@ -166,8 +172,10 @@ def search_for_similar_published_fact_checks_no_cache(type, query, team_ids, aft
         threshold = Bot::Alegre.get_threshold_for_query(type, pm)[0][:value]
-        alegre_results = Bot::Alegre.get_items_with_similar_media_v2(media_url, [{ value: threshold }], team_ids, type)
+        alegre_results = Bot::Alegre.get_items_with_similar_media_v2(media_url: media_url, threshold: [{ value: threshold }], team_ids: team_ids, type: type)
+        Rails.logger.error "ALEGRE RESULTS ARE #{alegre_results.inspect}"
```
This doesn't seem to be an error... if we still want to log this situation, let's log as `info` and use our logging pattern (`[Smooch Bot] <Message>`).

caiosba commented on app/models/concerns/smooch_search.rb:
```diff
@@ -166,8 +172,10 @@ def search_for_similar_published_fact_checks_no_cache(type, query, team_ids, aft
+        alegre_results = Bot::Alegre.get_items_with_similar_media_v2(media_url: media_url, threshold: [{ value: threshold }], team_ids: team_ids, type: type)
+        Rails.logger.error "ALEGRE RESULTS ARE #{alegre_results.inspect}"
         results = self.parse_search_results_from_alegre(alegre_results, after, feed_id, team_ids)
+        Rails.logger.error "RESULTS ARE #{results.inspect}"
```
This doesn't seem to be an error... if we still want to log this situation, let's log as `info` and use our logging pattern (`[Smooch Bot] <Message>`).

caiosba reviewed (changes requested):
Thanks Devin, I left some comments, some of which are only suggestions. Now I'm going to test it locally.

caiosba commented:
Devin, once you address the comments from my previous review, please re-request a final review. I finished testing it locally and I was able to get text and images to match for items created inside the application and also for tipline searches and feed queries:

![Captura de tela de 2024-04-06 19-09-47](https://github.com/meedan/check-api/assets/117518/6fb37518-a667-4f3e-955a-9373f54e608b)

![Captura de tela de 2024-04-06 19-10-18](https://github.com/meedan/check-api/assets/117518/8f93bf86-657f-4595-ab73-643e478ac848)

:tada:

DGaffney commented:
@caiosba I've reviewed your notes over the weekend and fixed them all! Take another look when you're ready :)

caiosba reviewed (changes requested):
Thanks @DGaffney ! There is one last one here: https://github.com/meedan/check-api/pull/1827#discussion_r1554431665

DGaffney commented:
> Thanks @DGaffney ! There is one last one here: [#1827 (comment)](https://github.com/meedan/check-api/pull/1827#discussion_r1554431665)

Fixed and updated!

caiosba reviewed (approved):
Thanks Devin for addressing all PR comments!

caiosba commented on app/models/concerns/alegre_v2.rb:
```diff
@@ -263,26 +327,138 @@ def get_confirmed_items(project_media, field)
+        while start_time + timeout > Time.now && cached_data.values.include?([])
+          sleep(1)
+          cached_data = get_cached_data(get_required_keys(project_media, nil))
+        end
```
@DGaffney can you please notify Sentry if we reach the timeout? This is how you can do it: 

(1) Define a specific class for this error, something like (at the top of the file):

```
class AlegreTimeoutError < StandardError; end
```

(2) Call Sentry:

```
CheckSentry.notify(AlegreTimeoutError.new('Timeout when waiting for async response from Alegre'), params: { project_media: project_media })
```

caiosba reviewed (approved):
Devin and I exchanged a few DMs about it - good to go!

caiosba commented:
*Good to go if tests pass, CC is happy, etc.
